services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt

```diff
@@ -13,36 +14,60 @@ class HankeKayttajaService(
 
     @Transactional
     fun saveNewTokensFromApplication(applicationData: ApplicationData, hankeId: Int) {
-        // Contacts new for this application
-        val newContacts =
+        val contacts =
             applicationData
                 .customersWithContacts()
                 .flatMap { it.contacts }
-                .filter { !it.email.isNullOrBlank() && !it.name.isNullOrBlank() }
-
-        val existingEmails =
-            hankeKayttajaRepository
-                .findByHankeIdAndSahkopostiIn(hankeId, newContacts.map { it.email!! })
-                .map { it.sahkoposti }
-
-        val contactsToCreate =
-            newContacts
-                .filter { contact -> !existingEmails.contains(contact.email) }
-                .distinctBy { it.email }
-        contactsToCreate.forEach { createToken(hankeId, it) }
+                .filterNot { it.email.isNullOrBlank() || it.name.isNullOrBlank() }
+                .map { UserContact(it.name!!, it.email!!) }
+
+        filterNewContacts(hankeId, contacts).forEach { contact -> createToken(hankeId, contact) }
+    }
+
+    @Transactional
+    fun saveNewTokensFromHanke(hanke: Hanke) {
+        val hankeId = hanke.id ?: throw HankeArgumentException("Hanke without id")
+        val contacts =
+            hanke
+                .extractYhteystiedot()
+                .flatMap { it.alikontaktit }
+                .mapNotNull { person ->
+                    val name = person.wholeName()
+                    when {
+                        name.isBlank() || person.email.isBlank() -> null
+                        else -> UserContact(name, person.email)
+                    }
+                }
```

Comment:
This can be flattened as well
```suggestion
                .filterNot { it.email.isBlank() || it.wholeName().isBlank() }
                .map { UserContact(it.wholeName(), it.email) }
```
If you're worried about calling `wholeName()` twice, it's a very cheap method. Or you could make it a lazy value.

Reply:
Fixed.